researchstudio-sat/webofneeds
# Fix groupfacet

How to test:
1. Run the grouping bot with the code before this PR. Observe that it does not work.
2. Run the grouping bot with the code in this PR. Observe that it creates 5 needs plus 1 group need, connects the needs to the group need, sends 5 messages from each need to the group, waits for all needs to receive 20 messages each (5 from each other group member), and then finishes.

Changes were made to message processing that should have left the rest of the WoN functionality unchanged.

* Fixed errors in the facet code

* Changed WonMessageRoutes to support the way the group facet works: each message 
that the group need receives on the GroupFacet is *forwarded*: a FromSystem message is
generated that has the complete content of the original message in its dataset. 
From its envelope, the msg:hasForwardedMessage property points to the URI of the 
original message. The new message is put into the SystemMessageIn route. 

* Added the possibility for a messageProcessor to suppress sending a message to the owner, 
and to react to the message 

* Changes in the WonMessage and WonMessageBuilder were necessary to implement 
forwarding correctly

* A new RDF property msg:hasForwardedMessage was added

* Message structure checks (the rq files) were adapted to admit forwarded messages

* A new processor was added for letting 'System' send a regular connection message

* A new predicate was added to check if a message is a System message destined at the 
owner, so we can distinguish between such messages (this is the case for close messages
caused by a deactivate, and by responses to owner-sent messages) and outbound System
messages, as it is the case in the GroupFacet implementation
